linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/JSGen.scala

```diff
@@ -350,15 +352,37 @@ private[emitter] final class JSGen(val semantics: Semantics,
 
   def genClassDataOf(typeRef: TypeRef)(implicit pos: Position): Tree = {
     typeRef match {
+      case typeRef: PrimRef =>
+        genClassDataOf(fieldNameOfPrimRef(typeRef))
       case ClassRef(className) =>
         genClassDataOf(className)
       case ArrayTypeRef(base, dims) =>
-        (1 to dims).foldLeft[Tree](envField("d", base)) { (prev, _) =>
+        val baseData = envField("d", arrayBaseFieldName(base))
+        (1 to dims).foldLeft[Tree](baseData) { (prev, _) =>
           Apply(DotSelect(prev, Ident("getArrayOf")), Nil)
         }
     }
   }
 
+  private def arrayBaseFieldName(base: NonArrayTypeRef): String = base match {
+    case base: PrimRef       => fieldNameOfPrimRef(base)
+    case ClassRef(className) => className
+  }
+
+  private def fieldNameOfPrimRef(primRef: PrimRef): String = (primRef: @unchecked) match {
+    case VoidRef    => "V"
+    case BooleanRef => "Z"
+    case CharRef    => "C"
+    case ByteRef    => "B"
+    case ShortRef   => "S"
+    case IntRef     => "I"
+    case LongRef    => "J"
+    case FloatRef   => "F"
+    case DoubleRef  => "D"
+    case NullRef    => "N"
+    case NothingRef => "E"
+  }
```

Comment:
I find it a bit strange that we introduce types but the emitter internally still uses char codes in many locations (n.b. the corejslib). Consider using this function everywhere to translate.